### ISSUE TITLE
Bug: between 1.10.35 and 1.10.44, ldap error

### ISSUE THREAD

**Issue by ianmiell** [IshentRas/cookbook-openshift3#115]
old snip:

```
  identityProviders:
  - name: ldap_identity
    challenge: true
    login: true
    provider:
      apiVersion: v1
      kind: BasicAuthPasswordIdentityProvider
      url: https://auth-service-openshift-infra.appcloud-pilot.barcapint.com/auth/ldap_auth
  masterPublicURL: 
```

new snip:

```
  - name: ldap_identity
    challenge: true
    login: true
    provider:
      apiVersion: v1
      kind: BasicAuthPasswordIdentityProvider
      url: https://auth-service-openshift-infra.appcloud-pilot.barcapint.com/auth/ldap_auth
        ca:
        certFile:
        keyFile:
  masterPublicURL: 
```

error: 

```
Apr 25 16:26:19 gdcdsr000002532.intranet.barcapint.com atomic-openshift-master-api[102923]: F0425 16:26:19.721911  102923 start_api.go:65] could not load config file "/etc/origin/master/master-config.yaml" due to an error: yaml: line 105: mapping values are not allowed in this context
```

**Comment by ianmiell** [comment]
I think the problem is that there was a change here:

```
$ git diff 58cd24f72ac427286b1929fb741ecbed0d30cf9d oauthConfig.erb
[...]
-<% for key in ['ca' , 'certFile' , 'keyFile'] %>
-<% if node['cookbook-openshift3']['openshift_master_identity_provider'][node['cookbook-openshift3']['oauth_Identity']].has_key?(key) %>
-     <%= key %> : <%= node['cookbook-openshift3']['openshift_master_identity_provider'][node['cookbook-openshift3']['oauth_Identity']][key] %>
-<% end %>
-<% end %>
-<% elsif node['cookbook-openshift3']['openshift_master_identity_provider'][node['cookbook-openshift3']['oauth_Identity']]['kind'] == 'RequestHeaderIdentityProvider' -%>
+      kind: <%= node['cookbook-openshift3']['openshift_master_identity_provider'][identityProvider]['kind'] %>
+  <% if node['cookbook-openshift3']['openshift_master_identity_provider'][identityProvider]['kind'] == 'HTPasswdPasswordIdentityProvider' -%>
+      file: <%= node['cookbook-openshift3']['openshift_master_identity_provider'][identityProvider]['filename'] %>
+  <% elsif node['cookbook-openshift3']['openshift_master_identity_provider'][identityProvider]['kind'] == 'BasicAuthPasswordIdentityProvider' -%>
+      url: <%= node['cookbook-openshift3']['openshift_master_identity_provider'][identityProvider]['url'] %>
+    <% ['ca' , 'certFile' , 'keyFile'].each do |key| %>
+      <% if node['cookbook-openshift3']['openshift_master_identity_provider'][identityProvider].has_key?(key) %>
+        <%= key %>: <%= node['cookbook-openshift3']['openshift_master_identity_provider'][identityProvider][key] %>
+      <% end %>
+    <% end %>
+  <% elsif node['cookbook-openshift3']['openshift_master_identity_provider'][identityProvider]['kind'] == 'RequestHeaderIdentityProvider' -%>
```

An assumption of the change was that the id provider would have certFile, keyFile. However, reading the code it looks like it only adds the key if it has one - so perhaps it's a change to a config item that defaults to a value.

**Comment by ianmiell** [comment]
Bug looks to be in our config - the format changed, and the defaults die. Which I believe is a feature.

**Comment by ianmiell** [comment]
OK,

**Comment by ianmiell** [comment]
https://coderanger.net/arrays-and-chef/

**Comment by ianmiell** [comment]
'However things are not as simple when the two values are at the same level. The two arrays are set-unioned together, with the resulting order being deterministic but difficult to predict to the casual reader ... Needless to say this can result in much confusion.'

**Comment by jperville** [comment]
Yes, chef attribute merging is very tricky, especially for arrays.

**Comment by ianmiell** [comment]
Fix has gone into v1.10.45
Ideally I'd like to create some test cases, but I need to do some work to prevent the time to test from being prohibitive.
